packages/dds/tree/api-report/tree.public.api.md

```diff
@@ -514,10 +525,11 @@ export interface TreeNodeSchemaCore<out Name extends string, out Kind extends No
     readonly info: Info;
     // (undocumented)
     readonly kind: Kind;
+    readonly metadata?: RecursiveReadonly<TMetadata> | undefined;
```

Comment:
I personally don't see any harm in custom metadata being mutable (I don't recommend it, but it shouldn't break anything). This RecursiveReadonly application does however break things, since it stripped private and protected fields as well as constructors, thus removing nominal typing and making types as listed in error and IntelliSense much more complex when they might have been simple classes before. 

Also given that readonly that the input value provided as the metadata could be modified by the code that passed it in after passing it in (since we don't deep copy: at least I assume we don't) this doesn't actually prevent mutation.

Also, given that typescript doesn't prevent assigning readonly values to writable ones, using readonly like this isn't too likely to actually stop anything.


Also this won't help when mutation is done through methods. For example this will not convert Map to ReadonlyMap, it will instead convert Map but expressed in a hard to read way.

Lastly, I see valid use-cases for mutable metadata. For example putting a registry of behaviors onto a schema could be useful, and this readonlyness doesn't actually manage to prevent you from doing it and mutating via this API anyway if you use a map.

Reply:
Ah, fair enough. I'll revert that aspect then.